File: cpp/speed_comp_python/fibonacci.cc
#include <iostream>
#include <chrono>

int fibonacci_recursive(int n) {
    if (n <= 0)
        return 0;
    else if (n == 1)
        return 1;
    else
        return fibonacci_recursive(n-1) + fibonacci_recursive(n-2);
}

int fibonacci_iterative(int n) {
    if (n <= 0)
        return 0;
    else if (n == 1)
        return 1;
    else {
        int a = 0, b = 1;
        for (int i = 2; i <= n; ++i) {
            int temp = b;
            b = a + b;
            a = temp;
        }
        return b;
    }
}

int main() {
    int n = 30;
    auto start_time = std::chrono::high_resolution_clock::now();
    int result_recursive = fibonacci_recursive(n);
    auto end_time = std::chrono::high_resolution_clock::now();
    std::chrono::duration<double> recursive_duration = end_time - start_time;
    std::cout << "Fibonacci recursive result: " << result_recursive << ", Time taken: " << recursive_duration.count() << " seconds" << std::endl;

    start_time = std::chrono::high_resolution_clock::now();
    int result_iterative = fibonacci_iterative(n);
    end_time = std::chrono::high_resolution_clock::now();
    std::chrono::duration<double> iterative_duration = end_time - start_time;
    std::cout << "Fibonacci iterative result: " << result_iterative << ", Time taken: " << iterative_duration.count() << " seconds" << std::endl;

    return 0;
}

File: cpp/speed_comp_python/fibonacci.py
import time

def fibonacci_recursive(n):
    if n <= 0:
        return 0
    elif n == 1:
        return 1
    else:
        return fibonacci_recursive(n-1) + fibonacci_recursive(n-2)


def fibonacci_iterative(n):
    if n <= 0:
        return 0
    elif n == 1:
        return 1
    else:
        a, b = 0, 1
        for _ in range(2, n+1):
            a, b = b, a + b
        return b

def main():
    n = 30
    start_time = time.time()
    result = fibonacci_recursive(n)
    end_time = time.time()
    print(f"Fibonacci recursive result: {result}, Time taken: {end_time - start_time} seconds")
    
    start_time = time.time()
    result = fibonacci_iterative(n)
    end_time = time.time()
    print(f"Fibonacci iterative result: {result}, Time taken: {end_time - start_time} seconds")

if __name__ == "__main__":
    main()
